File: app/src/main/java/me/zhanghai/android/files/provider/archive/archiver/ArchiveReader.java
/*
 * Copyright (c) 2018 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.provider.archive.archiver;

import android.content.Context;
import android.content.SharedPreferences;
import android.os.Build;
import android.util.Pair;

import org.apache.commons.compress.archivers.ArchiveEntry;
import org.apache.commons.compress.archivers.ArchiveInputStream;
import org.apache.commons.compress.archivers.ArchiveStreamFactory;
import org.apache.commons.compress.archivers.sevenz.SevenZArchiveEntry;
import org.apache.commons.compress.archivers.sevenz.SevenZFile;
import org.apache.commons.compress.archivers.tar.TarArchiveEntry;
import org.apache.commons.compress.archivers.zip.ZipArchiveEntry;
import org.apache.commons.compress.compressors.CompressorException;
import org.apache.commons.compress.compressors.CompressorStreamFactory;

import java.io.BufferedInputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;

import androidx.annotation.NonNull;
import androidx.preference.PreferenceManager;
import eu.chainfire.librootjava.RootJava;
import java8.nio.charset.StandardCharsets;
import java8.nio.file.AccessMode;
import java8.nio.file.NoSuchFileException;
import java8.nio.file.NotLinkException;
import java8.nio.file.Path;
import java9.lang.Iterables;
import java9.util.Iterators;
import me.zhanghai.android.files.BuildConfig;
import me.zhanghai.android.files.R;
import me.zhanghai.android.files.compat.EnumerationCompat;
import me.zhanghai.android.files.compat.MapCompat;
import me.zhanghai.android.files.provider.common.IsDirectoryException;
import me.zhanghai.android.files.provider.common.MoreFiles;
import me.zhanghai.android.files.provider.common.PosixFileType;
import me.zhanghai.android.files.provider.common.PosixFileTypes;
import me.zhanghai.android.files.provider.root.RootUtils;
import me.zhanghai.android.files.settings.SettingsLiveDatas;
import me.zhanghai.android.files.util.IoUtils;

public class ArchiveReader {

    @NonNull
    private static final CompressorStreamFactory sCompressorStreamFactory =
            new CompressorStreamFactory();
    @NonNull
    private static final ArchiveStreamFactory sArchiveStreamFactory = new ArchiveStreamFactory();

    private ArchiveReader() {}

    @NonNull
    public static Pair<Map<Path, ArchiveEntry>, Map<Path, List<Path>>> readEntries(
            @NonNull Path file, @NonNull Path rootPath) throws IOException {
        Map<Path, ArchiveEntry> entries = new HashMap<>();
        List<ArchiveEntry> rawEntries = readEntries(file);
        for (ArchiveEntry entry : rawEntries) {
            Path path = rootPath.resolve(entry.getName());
            // Normalize an absolute path to prevent path traversal attack.
            if (!path.isAbsolute()) {
                throw new AssertionError("Path must be absolute: " + path.toString());
            }
            if (path.getNameCount() > 0) {
                path = path.normalize();
                if (path.getNameCount() == 0) {
                    // Don't allow a path to become the root path only after normalization.
                    continue;
                }
            }
            MapCompat.putIfAbsent(entries, path, entry);
        }
        if (!entries.containsKey(rootPath)) {
            entries.put(rootPath, new DirectoryArchiveEntry(""));
        }
        Map<Path, List<Path>> tree = new HashMap<>();
        tree.put(rootPath, new ArrayList<>());
        List<Path> paths = new ArrayList<>(entries.keySet());
        for (Path path : paths) {
            while (true) {
                Path parentPath = path.getParent();
                if (parentPath == null) {
                    break;
                }
                ArchiveEntry entry = entries.get(path);
                if (entry.isDirectory()) {
                    MapCompat.computeIfAbsent(tree, path, _1 -> new ArrayList<>());
                }
                MapCompat.computeIfAbsent(tree, parentPath, _1 -> new ArrayList<>())
                        .add(path);
                if (entries.containsKey(parentPath)) {
                    break;
                }
                entries.put(parentPath, new DirectoryArchiveEntry(parentPath.toString()));
                path = parentPath;
            }
        }
        return new Pair<>(entries, tree);
    }

    @NonNull
    private static List<ArchiveEntry> readEntries(@NonNull Path file) throws IOException {
        File ioFile = file.toFile();
        String compressorType;
        String archiveType;
        try (FileInputStream fileInputStream = new FileInputStream(ioFile);
             BufferedInputStream bufferedInputStream = new BufferedInputStream(fileInputStream)) {
            try {
                compressorType = CompressorStreamFactory.detect(bufferedInputStream);
            } catch (CompressorException e) {
                // Ignored.
                compressorType = null;
            }
            try (BufferedInputStream bufferedCompressorInputStream = compressorType != null ?
                    new BufferedInputStream(sCompressorStreamFactory.createCompressorInputStream(
                            compressorType, bufferedInputStream)) : bufferedInputStream) {
                archiveType = ArchiveStreamFactory.detect(bufferedCompressorInputStream);
            } catch (org.apache.commons.compress.archivers.ArchiveException e) {
                throw new ArchiveException(e);
            } catch (CompressorException e) {
                throw new ArchiveException(e);
            }
        } catch (FileNotFoundException e) {
            MoreFiles.checkAccess(file, AccessMode.READ);
            NoSuchFileException noSuchFileException = new NoSuchFileException(file.toString());
            noSuchFileException.initCause(e);
            throw noSuchFileException;
        }
        String encoding = getArchiveFileNameEncoding();
        ArrayList<ArchiveEntry> entries = new ArrayList<>();
        if (compressorType == null) {
            switch (archiveType) {
                case ArchiveStreamFactory.ZIP:
                    try (ZipFileCompat zipFile = new ZipFileCompat(ioFile, encoding)) {
                        Iterators.forEachRemaining(EnumerationCompat.asIterator(
                                zipFile.getEntries()), entries::add);
                        return entries;
                    }
                case ArchiveStreamFactory.SEVEN_Z:
                    if (Build.VERSION.SDK_INT < Build.VERSION_CODES.N) {
                        throw new IOException(new UnsupportedOperationException("SevenZFile"));
                    }
                    try (SevenZFile sevenZFile = new SevenZFile(ioFile)) {
                        Iterables.forEach(sevenZFile.getEntries(), entries::add);
                        return entries;
                    }
            }
        }
        try (FileInputStream fileInputStream = new FileInputStream(ioFile);
             BufferedInputStream bufferedInputStream = new BufferedInputStream(fileInputStream);
             InputStream compressorInputStream = compressorType != null ?
                     sCompressorStreamFactory.createCompressorInputStream(compressorType,
                             bufferedInputStream) : bufferedInputStream;
             ArchiveInputStream archiveInputStream = sArchiveStreamFactory.createArchiveInputStream(
                     archiveType, compressorInputStream, encoding)) {
            ArchiveEntry entry;
            while ((entry = archiveInputStream.getNextEntry()) != null) {
                entries.add(entry);
            }
            return entries;
        } catch (FileNotFoundException e) {
            NoSuchFileException noSuchFileException = new NoSuchFileException(
                    file.toString());
            noSuchFileException.initCause(e);
            throw noSuchFileException;
        } catch (org.apache.commons.compress.archivers.ArchiveException e) {
            throw new ArchiveException(e);
        } catch (CompressorException e) {
            throw new ArchiveException(e);
        }
    }

    @NonNull
    private static String getArchiveFileNameEncoding() {
        if (RootUtils.isRunningAsRoot()) {
            try {
                Context context = RootJava.getPackageContext(BuildConfig.APPLICATION_ID);
                SharedPreferences sharedPreferences = PreferenceManager.getDefaultSharedPreferences(
                        context);
                String key = context.getString(R.string.pref_key_archive_file_name_encoding);
                String defaultValue = context.getString(
                        R.string.pref_default_value_archive_file_name_encoding);
                return sharedPreferences.getString(key, defaultValue);
            } catch (Exception e) {
                e.printStackTrace();
                return StandardCharsets.UTF_8.name();
            }
        } else {
            return SettingsLiveDatas.ARCHIVE_FILE_NAME_ENCODING.getValue();
        }
    }

    @NonNull
    public static InputStream newInputStream(@NonNull Path file, @NonNull ArchiveEntry entry)
            throws IOException {
        if (entry.isDirectory()) {
            throw new IsDirectoryException(file.toString());
        }
        File ioFile = file.toFile();
        String compressorType;
        String archiveType;
        try (FileInputStream fileInputStream = new FileInputStream(ioFile);
             BufferedInputStream bufferedInputStream = new BufferedInputStream(fileInputStream)) {
            try {
                compressorType = CompressorStreamFactory.detect(bufferedInputStream);
            } catch (CompressorException e) {
                // Ignored.
                compressorType = null;
            }
            try (BufferedInputStream bufferedCompressorInputStream = compressorType != null ?
                    new BufferedInputStream(sCompressorStreamFactory.createCompressorInputStream(
                            compressorType, bufferedInputStream)) : bufferedInputStream) {
                archiveType = ArchiveStreamFactory.detect(bufferedCompressorInputStream);
            } catch (org.apache.commons.compress.archivers.ArchiveException e) {
                throw new ArchiveException(e);
            } catch (CompressorException e) {
                throw new ArchiveException(e);
            }
        } catch (FileNotFoundException e) {
            MoreFiles.checkAccess(file, AccessMode.READ);
            NoSuchFileException noSuchFileException = new NoSuchFileException(file.toString());
            noSuchFileException.initCause(e);
            throw noSuchFileException;
        }
        String encoding = getArchiveFileNameEncoding();
        if (compressorType == null) {
            if (entry instanceof ZipArchiveEntry) {
                ZipArchiveEntry zipEntry = (ZipArchiveEntry) entry;
                boolean successful = false;
                ZipFileCompat zipFile = null;
                InputStream zipEntryInputStream = null;
                try {
                    zipFile = new ZipFileCompat(ioFile, encoding);
                    zipEntryInputStream = zipFile.getInputStream(zipEntry);
                    if (zipEntryInputStream == null) {
                        throw new NoSuchFileException(file.toString());
                    }
                    InputStream inputStream = new ZipFileEntryInputStream(zipFile,
                            zipEntryInputStream);
                    successful = true;
                    return inputStream;
                } finally {
                    if (!successful) {
                        if (zipEntryInputStream != null) {
                            zipEntryInputStream.close();
                        }
                        if (zipFile != null) {
                            zipFile.close();
                        }
                    }
                }
            } else if (entry instanceof SevenZArchiveEntry) {
                boolean successful = false;
                SevenZFile sevenZFile = null;
                try {
                    sevenZFile = new SevenZFile(ioFile);
                    SevenZArchiveEntry currentEntry;
                    while ((currentEntry = sevenZFile.getNextEntry()) != null) {
                        if (!Objects.equals(currentEntry.getName(), entry.getName())) {
                            continue;
                        }
                        InputStream inputStream = new SevenZArchiveEntryInputStream(sevenZFile,
                                currentEntry);
                        successful = true;
                        return inputStream;
                    }
                    throw new NoSuchFileException(file.toString());
                } finally {
                    if (!successful && sevenZFile != null) {
                        sevenZFile.close();
                    }
                }
            }
        }
        boolean successful = false;
        FileInputStream fileInputStream = null;
        BufferedInputStream bufferedInputStream = null;
        InputStream compressorInputStream = null;
        ArchiveInputStream archiveInputStream = null;
        try {
            fileInputStream = new FileInputStream(ioFile);
            bufferedInputStream = new BufferedInputStream(fileInputStream);
            compressorInputStream = compressorType != null ?
                    sCompressorStreamFactory.createCompressorInputStream(compressorType,
                            bufferedInputStream) : bufferedInputStream;
            archiveInputStream = sArchiveStreamFactory.createArchiveInputStream(archiveType,
                    compressorInputStream, encoding);
            ArchiveEntry currentEntry;
            while ((currentEntry = archiveInputStream.getNextEntry()) != null) {
                if (!Objects.equals(currentEntry.getName(), entry.getName())) {
                    continue;
                }
                successful = true;
                return archiveInputStream;
            }
            throw new NoSuchFileException(file.toString());
        } catch (FileNotFoundException e) {
            NoSuchFileException noSuchFileException = new NoSuchFileException(
                    file.toString());
            noSuchFileException.initCause(e);
            throw noSuchFileException;
        } catch (org.apache.commons.compress.archivers.ArchiveException e) {
            throw new ArchiveException(e);
        } catch (CompressorException e) {
            throw new ArchiveException(e);
        } finally {
            if (!successful) {
                if (archiveInputStream != null) {
                    archiveInputStream.close();
                }
                if (compressorInputStream != null) {
                    compressorInputStream.close();
                }
                if (bufferedInputStream != null) {
                    bufferedInputStream.close();
                }
                if (fileInputStream != null) {
                    fileInputStream.close();
                }
            }
        }
    }

    @NonNull
    public static String readSymbolicLink(@NonNull Path file, @NonNull ArchiveEntry entry)
            throws IOException {
        if (!isSymbolicLink(entry)) {
            throw new NotLinkException(file.toString());
        }
        if (entry instanceof TarArchiveEntry) {
            TarArchiveEntry tarEntry = (TarArchiveEntry) entry;
            return tarEntry.getLinkName();
        } else {
            try (InputStream inputStream = newInputStream(file, entry)) {
                return IoUtils.inputStreamToString(inputStream, StandardCharsets.UTF_8);
            }
        }
    }

    private static boolean isSymbolicLink(@NonNull ArchiveEntry entry) {
        return PosixFileTypes.fromArchiveEntry(entry) == PosixFileType.SYMBOLIC_LINK;
    }

    private static class DirectoryArchiveEntry implements ArchiveEntry {

        @NonNull
        private String mName;

        public DirectoryArchiveEntry(@NonNull String name) {
            if (name.endsWith("/")) {
                throw new IllegalArgumentException("name should not end with a slash: " + name);
            }
            mName = name + "/";
        }

        @NonNull
        @Override
        public String getName() {
            return mName;
        }

        @Override
        public long getSize() {
            return 0;
        }

        @Override
        public boolean isDirectory() {
            return true;
        }

        @NonNull
        @Override
        public Date getLastModifiedDate() {
            return new Date(-1);
        }

        @Override
        public boolean equals(Object object) {
            if (this == object) {
                return true;
            }
            if (object == null || getClass() != object.getClass()) {
                return false;
            }
            DirectoryArchiveEntry that = (DirectoryArchiveEntry) object;
            return Objects.equals(mName, that.mName);
        }

        @Override
        public int hashCode() {
            return Objects.hash(mName);
        }
    }

    private static class ZipFileEntryInputStream extends InputStream {

        @NonNull
        private final ZipFileCompat mFile;
        @NonNull
        private final InputStream mInputStream;

        public ZipFileEntryInputStream(@NonNull ZipFileCompat file,
                                       @NonNull InputStream inputStream) {
            mFile = file;
            mInputStream = inputStream;
        }

        @Override
        public int available() throws IOException {
            return mInputStream.available();
        }

        @Override
        public int read() throws IOException {
            return mInputStream.read();
        }

        @Override
        public int read(@NonNull byte[] b) throws IOException {
            return mInputStream.read(b);
        }

        @Override
        public int read(@NonNull byte[] b, int off, int len) throws IOException {
            return mInputStream.read(b, off, len);
        }

        @Override
        public void close() throws IOException {
            mInputStream.close();
            mFile.close();
        }
    }

    private static class SevenZArchiveEntryInputStream extends InputStream {

        @NonNull
        private final SevenZFile mFile;
        @NonNull
        private final SevenZArchiveEntry mEntry;

        public SevenZArchiveEntryInputStream(@NonNull SevenZFile file,
                                             @NonNull SevenZArchiveEntry entry) {
            mFile = file;
            mEntry = entry;
        }

        @Override
        public int available() {
            long size = mEntry.getSize();
            long read = mFile.getStatisticsForCurrentEntry()
                    .getUncompressedCount();
            long available = size - read;
            return (int) Math.min(available, Integer.MAX_VALUE);
        }

        @Override
        public int read() throws IOException {
            return mFile.read();
        }

        @Override
        public int read(@NonNull byte[] b) throws IOException {
            return mFile.read(b);
        }

        @Override
        public int read(@NonNull byte[] b, int off, int len) throws IOException {
            return mFile.read(b, off, len);
        }

        @Override
        public void close() throws IOException {
            mFile.close();
        }
    }
}

File: app/src/main/java/me/zhanghai/android/files/provider/archive/archiver/ArchiveReader.kt
/*
 * Copyright (c) 2018 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.provider.archive.archiver

import androidx.preference.PreferenceManager
import java8.nio.channels.SeekableByteChannel
import java8.nio.charset.StandardCharsets
import java8.nio.file.Path
import me.zhanghai.android.files.R
import me.zhanghai.android.files.provider.common.DelegateForceableSeekableByteChannel
import me.zhanghai.android.files.provider.common.DelegateInputStream
import me.zhanghai.android.files.provider.common.DelegateNonForceableSeekableByteChannel
import me.zhanghai.android.files.provider.common.ForceableChannel
import me.zhanghai.android.files.provider.common.PosixFileMode
import me.zhanghai.android.files.provider.common.PosixFileType
import me.zhanghai.android.files.provider.common.newByteChannel
import me.zhanghai.android.files.provider.common.newInputStream
import me.zhanghai.android.files.provider.root.isRunningAsRoot
import me.zhanghai.android.files.provider.root.rootContext
import me.zhanghai.android.files.settings.Settings
import me.zhanghai.android.files.util.valueCompat
import java.io.Closeable
import java.io.IOException
import java.io.InputStream
import java.nio.charset.Charset

object ArchiveReader {
    @Throws(IOException::class)
    fun readEntries(
        file: Path,
        passwords: List<String>,
        rootPath: Path
    ): Pair<Map<Path, ReadArchive.Entry>, Map<Path, List<Path>>> {
        val entries = mutableMapOf<Path, ReadArchive.Entry>()
        val rawEntries = readEntries(file, passwords)
        for (entry in rawEntries) {
            var path = rootPath.resolve(entry.name)
            // Normalize an absolute path to prevent path traversal attack.
            if (!path.isAbsolute) {
                // TODO: Will this actually happen?
                throw AssertionError("Path must be absolute: $path")
            }
            if (path.nameCount > 0) {
                path = path.normalize()
                if (path.nameCount == 0) {
                    // Don't allow a path to become the root path only after normalization.
                    continue
                }
            } else {
                if (!entry.isDirectory) {
                    // Ignore a root path that's not a directory
                    continue
                }
            }
            entries.getOrPut(path) { entry }
        }
        entries.getOrPut(rootPath) { createDirectoryEntry("") }
        val tree = mutableMapOf<Path, MutableList<Path>>()
        tree[rootPath] = mutableListOf()
        val paths = entries.keys.toList()
        for (path in paths) {
            var path = path
            while (true) {
                val parentPath = path.parent ?: break
                val entry = entries[path]!!
                if (entry.isDirectory) {
                    tree.getOrPut(path) { mutableListOf() }
                }
                tree.getOrPut(parentPath) { mutableListOf() }.add(path)
                if (entries.containsKey(parentPath)) {
                    break
                }
                entries[parentPath] = createDirectoryEntry(parentPath.toString())
                path = parentPath
            }
        }
        return entries to tree
    }

    private fun createDirectoryEntry(name: String): ReadArchive.Entry {
        require(!name.endsWith("/")) { "name $name should not end with a slash" }
        return ReadArchive.Entry(
            name, false, null, null, null, PosixFileType.DIRECTORY, 0, null, null,
            PosixFileMode.DIRECTORY_DEFAULT, null
        )
    }

    @Throws(IOException::class)
    private fun readEntries(file: Path, passwords: List<String>): List<ReadArchive.Entry> {
        val charset = archiveFileNameCharset
        val (archive, closeable) = openArchive(file, passwords)
        return closeable.use {
            buildList {
                while (true) {
                    this += archive.readEntry(charset) ?: break
                }
            }
        }
    }

    @Throws(IOException::class)
    fun newInputStream(file: Path, passwords: List<String>, entry: ReadArchive.Entry): InputStream? {
        val charset = archiveFileNameCharset
        val (archive, closeable) = openArchive(file, passwords)
        var successful = false
        return try {
            while (true) {
                val currentEntry = archive.readEntry(charset) ?: break
                if (currentEntry.name != entry.name) {
                    continue
                }
                successful = true
                break
            }
            if (successful) {
                CloseableInputStream(archive.newDataInputStream(), closeable)
            } else {
                null
            }
        } finally {
            if (!successful) {
                closeable.close()
            }
        }
    }

    @Throws(IOException::class)
    private fun openArchive(
        file: Path,
        passwords: List<String>
    ): Pair<ReadArchive, ArchiveCloseable> {
        val channel = try {
            CacheSizeSeekableByteChannel(file.newByteChannel())
        } catch (e: Exception) {
            e.printStackTrace()
            null
        }
        if (channel != null) {
            var successful = false
            try {
                val archive = ReadArchive(channel, passwords)
                successful = true
                return archive to ArchiveCloseable(archive, channel)
            } finally {
                if (!successful) {
                    channel.close()
                }
            }
        }
        val inputStream = file.newInputStream()
        var successful = false
        try {
            val archive = ReadArchive(inputStream, passwords)
            successful = true
            return archive to ArchiveCloseable(archive, inputStream)
        } finally {
            if (!successful) {
                inputStream.close()
            }
        }
    }

    // size() may be called repeatedly for ZIP and 7Z, so make it cached to improve performance.
    private fun CacheSizeSeekableByteChannel(channel: SeekableByteChannel): SeekableByteChannel =
        if (channel is ForceableChannel) {
            CacheSizeForceableSeekableByteChannel(channel)
        } else {
            CacheSizeNonForceableSeekableByteChannel(channel)
        }

    private class CacheSizeNonForceableSeekableByteChannel(
        channel: SeekableByteChannel
    ) : DelegateNonForceableSeekableByteChannel(channel) {
        private val size: Long by lazy { super.size() }

        override fun size(): Long = size
    }

    private class CacheSizeForceableSeekableByteChannel(
        channel: SeekableByteChannel
    ) : DelegateForceableSeekableByteChannel(channel) {
        private val size: Long by lazy { super.size() }

        override fun size(): Long = size
    }

    private val archiveFileNameCharset: Charset
        get() =
            if (isRunningAsRoot) {
                try {
                    val sharedPreferences =
                        PreferenceManager.getDefaultSharedPreferences(rootContext)
                    val key = rootContext.getString(R.string.pref_key_archive_file_name_encoding)
                    val defaultValue = rootContext.getString(
                        R.string.pref_default_value_archive_file_name_encoding
                    )
                    Charset.forName(sharedPreferences.getString(key, defaultValue)!!)
                } catch (e: Exception) {
                    e.printStackTrace()
                    StandardCharsets.UTF_8
                }
            } else {
                Charset.forName(Settings.ARCHIVE_FILE_NAME_ENCODING.valueCompat)
            }

    private class ArchiveCloseable(
        private val archive: ReadArchive,
        private val closeable: Closeable
    ) : Closeable {
        override fun close() {
            @Suppress("ConvertTryFinallyToUseCall")
            try {
                archive.close()
            } finally {
                closeable.close()
            }
        }
    }

    private class CloseableInputStream(
        inputStream: InputStream,
        private val closeable: Closeable
    ) : DelegateInputStream(inputStream) {
        @Throws(IOException::class)
        override fun close() {
            super.close()

            closeable.close()
        }
    }
}
